crates/cext/src/transpiler/transpile_layout.rs

```diff
@@ -0,0 +1,507 @@
+    let layout = unsafe { const_ptr_as_ref(layout) };
+    let result = layout.final_index_layout(filter_ancillas);
+    // SAFETY: Per the documentation final_layout must be a valid pointer with a sufficient
+    // allocation for the output array
+    unsafe {
+        let out_slice = std::slice::from_raw_parts_mut(final_layout, result.len());
+        out_slice
+            .iter_mut()
+            .zip(result.iter())
+            .for_each(|(dest, src)| *dest = src.0);
+    }
+}
+
+/// @ingroup QkTranspileLayout
+/// Free a ``QkTranspileLayout`` object
+///
+/// @param layout a pointer to the layout to free
+///
+/// # Safety
+///
+/// Behavior is undefined if ``layout`` is not a valid, non-null pointer to a ``QkTranspileLayout``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_transpile_layout_free(layout: *mut TranspileLayout) {
+    if !layout.is_null() {
+        if !layout.is_aligned() {
+            panic!("Attempted to free a non-aligned pointer.")
+        }
+        // SAFETY: We have verified the pointer is non-null and aligned, so
+        // it should be readable by Box.
+        unsafe {
+            let _ = Box::from_raw(layout);
+        }
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use qiskit_circuit::bit::ShareableQubit;
+    use qiskit_circuit::nlayout::{NLayout, PhysicalQubit};
+    use qiskit_circuit::Qubit;
+    use qiskit_transpiler::transpile_layout::TranspileLayout;
+
+    #[test]
+    fn test_final_layout() {
+        let initial_layout_vec = vec![
+            PhysicalQubit(9),
+            PhysicalQubit(4),
+            PhysicalQubit(0),
+            PhysicalQubit(1),
+            PhysicalQubit(2),
+            PhysicalQubit(3),
+            PhysicalQubit(5),
+            PhysicalQubit(6),
+            PhysicalQubit(7),
+            PhysicalQubit(8),
+        ];
+        let initial_layout = NLayout::from_virtual_to_physical(initial_layout_vec).unwrap();
+        let routing_permutation = vec![
+            Qubit(2),
+            Qubit(0),
+            Qubit(1),
+            Qubit(4),
+            Qubit(5),
+            Qubit(6),
+            Qubit(7),
+            Qubit(8),
+            Qubit(9),
+            Qubit(3),
+        ];
+        let input_qubits = vec![ShareableQubit::new_anonymous(); 10];
+        let layout = TranspileLayout::new(
+            Some(initial_layout),
+            Some(routing_permutation),
+            input_qubits,
+            3,
+        );
+        let result = layout.final_index_layout(true);
+        let expected = vec![PhysicalQubit(3), PhysicalQubit(5), PhysicalQubit(2)];
+        assert_eq!(expected, result);
+        let mut result = vec![u32::MAX; layout.num_input_qubits() as usize];
+        unsafe { qk_transpile_layout_final_layout(&layout, true, result.as_mut_ptr()) };
+        let expected = expected.into_iter().map(|x| x.0).collect::<Vec<_>>();
+        assert_eq!(expected.as_slice(), result.as_slice());
```

Comment:
Nitpicky: it might have been a nice test to allocate the `Vec` longer than needed and then assert that we only wrote out the correct number of qubits.  But in theory, Miri should catch that anyway.